src/acl/external/time_quota/ext_time_quota_acl.cc

```diff
@@ -72,141 +68,106 @@ TDB_CONTEXT *db = nullptr;
 static int tq_debug_enabled = false;
 
 static void open_log(const char *logfilename)
 {
-    logfile = fopen(logfilename, "a");
-    if ( logfile == NULL ) {
+    static auto log = new std::ofstream(logfilename, std::ios::app);
+    if (log && *log)
+        logfile = log;
+    else
         perror(logfilename);
-        logfile = stderr;
-    }
-}
-
-static void vlog(const char *level, const char *format, va_list args)
-{
-    time_t now = time(NULL);
-
-    fprintf(logfile, "%ld %s| %s: ", static_cast<long int>(now),
-            program_name, level);
-    vfprintf (logfile, format, args);
-    fflush(logfile);
 }
 
-static void log_debug(const char *format, ...)
-{
-    va_list args;
-
-    if ( tq_debug_enabled ) {
-        va_start (args, format);
-        vlog("DEBUG", format, args);
-        va_end (args);
+#define log_debug(CONTENT) \
+    if (tq_debug_enabled) { \
+         *logfile << \
+            std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) << \
+             " DEBUG| " << CONTENT << std::endl ;\
     }
-}
 
-static void log_info(const char *format, ...)
-{
-    va_list args;
-
-    va_start (args, format);
-    vlog("INFO", format, args);
-    va_end (args);
-}
-
-static void log_error(const char *format, ...)
-{
-    va_list args;
+#define low_log(LEVEL, CONTENT) \
+    do { *logfile << \
+        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) << \
+        ' ' << LEVEL << "| " << CONTENT << std::endl ; } while (0)
 
-    va_start (args, format);
-    vlog("ERROR", format, args);
-    va_end (args);
-}
+#define log_info(CONTENT) low_log("INFO", CONTENT)
+#define log_error(CONTENT) low_log("ERROR", CONTENT)
+#define log_fatal(CONTENT) low_log("FATAL", CONTENT)
 
-static void log_fatal(const char *format, ...)
+static void init_db(void)
 {
-    va_list args;
+    log_info("opening time quota database \"" << db_path << "\".");
 
-    va_start (args, format);
-    vlog("FATAL", format, args);
-    va_end (args);
-}
+    int dbopts = 0;
+    if (noClearDb)
+        dbopts |= TDB_CLEAR_IF_FIRST;
 
-static void init_db(void)
-{
-    log_info("opening time quota database \"%s\".\n", db_path);
-    db = tdb_open(db_path, 0, TDB_CLEAR_IF_FIRST, O_CREAT | O_RDWR, 0666);
+    db = tdb_open(db_path, 0, dbopts, O_CREAT | O_RDWR, 0666);
     if (!db) {
-        log_fatal("Failed to open time_quota db '%s'\n", db_path);
+        log_fatal("Failed to open time_quota db '" << db_path << '\'');
         exit(EXIT_FAILURE);
     }
+    if (tq_debug_enabled) {
+        // count the number of entries in the database, only used for debugging
+        auto count = tdb_traverse(db,nullptr, nullptr);
+        log_debug("Database contains " << count << " entries.");
+    }
 }
 
 static void shutdown_db(void)
 {
     tdb_close(db);
 }
 
-static char *KeyString(int &len, const char *user_key, const char *sub_key)
+static std::string KeyString(const char *user_key, const char *sub_key)
```

Comment:
Please try to convert to SBuf instead of std::string. Using SBUf may reduce significant headaches with reusing Squid code. IIRC, recent primary code improvements allowed us to use SBuf in some other helpers.

Reply:
done